Add version argument

agila5 opened the pull request:
It can be used to download old OSM extracts from Geofabrik provider. Fixes #295. 

For example: 

``` r
library(osmextract)
#> Data (c) OpenStreetMap contributors, ODbL 1.0. https://www.openstreetmap.org/copyright.
#> Check the package website, https://docs.ropensci.org/osmextract/, for more details.
# Default behaviour
oe_get("Isle of Man", download_directory = tempdir())
#> The input place was matched with: Isle of Man
#> Downloading the OSM extract:
#>   |                                                                              |                                                                      |   0%  |                                                                              |                                                                      |   1%  |                                                                              |=                                                                     |   1%  |                                                                              |=                                                                     |   2%  |                                                                              |==                                                                    |   2%  |                                                                              |==                                                                    |   3%  |                                                                              |==                                                                    |   4%  |                                                                              |===                                                                   |   4%  |                                                                              |===                                                                   |   5%  |                                                                              |====                                                                  |   5%  |                                                                              |====                                                                  |   6%  |                                                                              |=====                                                                 |   6%  |                                                                              |=====                                                                 |   7%  |                                                                              |=====                                                                 |   8%  |                                                                              |======                                                                |   8%  |                                                                              |======                                                                |   9%  |                                                                              |=======                                                               |   9%  |                                                                              |=======                                                               |  10%  |                                                                              |=========                                                             |  13%  |                                                                              |==========                                                            |  14%  |                                                                              |==========                                                            |  15%  |                                                                              |===========                                                           |  15%  |                                                                              |===========                                                           |  16%  |                                                                              |============                                                          |  17%  |                                                                              |============                                                          |  18%  |                                                                              |=============                                                         |  18%  |                                                                              |================                                                      |  22%  |                                                                              |================                                                      |  23%  |                                                                              |================                                                      |  24%  |                                                                              |=================                                                     |  24%  |                                                                              |=================                                                     |  25%  |                                                                              |==================                                                    |  25%  |                                                                              |==================                                                    |  26%  |                                                                              |===================                                                   |  26%  |                                                                              |===================                                                   |  27%  |                                                                              |===================                                                   |  28%  |                                                                              |====================                                                  |  28%  |                                                                              |====================                                                  |  29%  |                                                                              |=====================                                                 |  29%  |                                                                              |=====================                                                 |  30%  |                                                                              |=====================                                                 |  31%  |                                                                              |======================                                                |  31%  |                                                                              |======================                                                |  32%  |                                                                              |=======================                                               |  32%  |                                                                              |=======================                                               |  33%  |                                                                              |=======================                                               |  34%  |                                                                              |========================                                              |  34%  |                                                                              |========================                                              |  35%  |                                                                              |=========================                                             |  35%  |                                                                              |=========================                                             |  36%  |                                                                              |==========================                                            |  36%  |                                                                              |==========================                                            |  37%  |                                                                              |==========================                                            |  38%  |                                                                              |===========================                                           |  38%  |                                                                              |===========================                                           |  39%  |                                                                              |============================                                          |  39%  |                                                                              |============================                                          |  40%  |                                                                              |============================                                          |  41%  |                                                                              |=============================                                         |  41%  |                                                                              |=============================                                         |  42%  |                                                                              |==============================                                        |  42%  |                                                                              |==============================                                        |  43%  |                                                                              |===============================                                       |  44%  |                                                                              |===============================                                       |  45%  |                                                                              |================================                                      |  45%  |                                                                              |================================                                      |  46%  |                                                                              |=================================                                     |  47%  |                                                                              |=================================                                     |  48%  |                                                                              |==================================                                    |  48%  |                                                                              |==================================                                    |  49%  |                                                                              |===================================                                   |  49%  |                                                                              |===================================                                   |  50%  |                                                                              |=====================================                                 |  53%  |                                                                              |=====================================                                 |  54%  |                                                                              |======================================                                |  54%  |                                                                              |======================================                                |  55%  |                                                                              |=======================================                               |  55%  |                                                                              |=======================================                               |  56%  |                                                                              |========================================                              |  56%  |                                                                              |========================================                              |  57%  |                                                                              |========================================                              |  58%  |                                                                              |=========================================                             |  58%  |                                                                              |=========================================                             |  59%  |                                                                              |==========================================                            |  61%  |                                                                              |===========================================                           |  61%  |                                                                              |===========================================                           |  62%  |                                                                              |============================================                          |  62%  |                                                                              |============================================                          |  63%  |                                                                              |============================================                          |  64%  |                                                                              |=============================================                         |  64%  |                                                                              |=============================================                         |  65%  |                                                                              |==============================================                        |  65%  |                                                                              |==============================================                        |  66%  |                                                                              |===============================================                       |  67%  |                                                                              |===============================================                       |  68%  |                                                                              |================================================                      |  68%  |                                                                              |================================================                      |  69%  |                                                                              |=================================================                     |  69%  |                                                                              |=================================================                     |  70%  |                                                                              |=================================================                     |  71%  |                                                                              |==================================================                    |  71%  |                                                                              |==================================================                    |  72%  |                                                                              |===================================================                   |  72%  |                                                                              |===================================================                   |  73%  |                                                                              |===================================================                   |  74%  |                                                                              |====================================================                  |  74%  |                                                                              |====================================================                  |  75%  |                                                                              |=====================================================                 |  75%  |                                                                              |=====================================================                 |  76%  |                                                                              |======================================================                |  76%  |                                                                              |======================================================                |  77%  |                                                                              |======================================================                |  78%  |                                                                              |=======================================================               |  78%  |                                                                              |=======================================================               |  79%  |                                                                              |========================================================              |  79%  |                                                                              |========================================================              |  80%  |                                                                              |========================================================              |  81%  |                                                                              |=========================================================             |  81%  |                                                                              |=========================================================             |  82%  |                                                                              |==========================================================            |  82%  |                                                                              |==========================================================            |  83%  |                                                                              |===========================================================           |  84%  |                                                                              |===========================================================           |  85%  |                                                                              |============================================================          |  85%  |                                                                              |============================================================          |  86%  |                                                                              |=============================================================         |  86%  |                                                                              |=============================================================         |  87%  |                                                                              |=============================================================         |  88%  |                                                                              |==============================================================        |  88%  |                                                                              |==============================================================        |  89%  |                                                                              |===============================================================       |  89%  |                                                                              |===============================================================       |  90%  |                                                                              |===============================================================       |  91%  |                                                                              |================================================================      |  91%  |                                                                              |================================================================      |  92%  |                                                                              |=================================================================     |  92%  |                                                                              |=================================================================     |  93%  |                                                                              |==================================================================    |  94%  |                                                                              |==================================================================    |  95%  |                                                                              |===================================================================   |  95%  |                                                                              |===================================================================   |  96%  |                                                                              |====================================================================  |  97%  |                                                                              |====================================================================  |  98%  |                                                                              |===================================================================== |  98%  |                                                                              |===================================================================== |  99%  |                                                                              |======================================================================|  99%  |                                                                              |======================================================================| 100%
#> File downloaded!
#> Starting with the vectortranslate operations on the input file!
#> 0...10...20...30...40...50...60...70...80...90...100 - done.
#> Finished the vectortranslate operations on the input file!
#> Reading layer `lines' from data source 
#>   `C:\Users\user\AppData\Local\Temp\RtmpyclUjo\geofabrik_isle-of-man-latest.gpkg' 
#>   using driver `GPKG'
#> Simple feature collection with 27379 features and 10 fields
#> Geometry type: LINESTRING
#> Dimension:     XY
#> Bounding box:  xmin: -20.05167 ymin: 48.18019 xmax: -2.913542 ymax: 55.71849
#> Geodetic CRS:  WGS 84

# An older file
oe_get("Isle of Man", download_directory = tempdir(), version = "140101")
#> The input place was matched with: Isle of Man
#> Downloading the OSM extract:
#>   |                                                                              |                                                                      |   0%  |                                                                              |                                                                      |   1%  |                                                                              |=                                                                     |   1%  |                                                                              |=                                                                     |   2%  |                                                                              |==                                                                    |   2%  |                                                                              |==                                                                    |   3%  |                                                                              |===                                                                   |   4%  |                                                                              |===                                                                   |   5%  |                                                                              |====                                                                  |   5%  |                                                                              |====                                                                  |   6%  |                                                                              |=====                                                                 |   7%  |                                                                              |=====                                                                 |   8%  |                                                                              |======                                                                |   8%  |                                                                              |======                                                                |   9%  |                                                                              |=======                                                               |   9%  |                                                                              |=======                                                               |  10%  |                                                                              |=======                                                               |  11%  |                                                                              |========                                                              |  11%  |                                                                              |========                                                              |  12%  |                                                                              |=========                                                             |  13%  |                                                                              |==========                                                            |  14%  |                                                                              |==========                                                            |  15%  |                                                                              |===========                                                           |  15%  |                                                                              |===========                                                           |  16%  |                                                                              |============                                                          |  17%  |                                                                              |============                                                          |  18%  |                                                                              |=============                                                         |  18%  |                                                                              |=============                                                         |  19%  |                                                                              |==============                                                        |  20%  |                                                                              |==============                                                        |  21%  |                                                                              |===============                                                       |  21%  |                                                                              |===============                                                       |  22%  |                                                                              |================                                                      |  22%  |                                                                              |================                                                      |  23%  |                                                                              |=================                                                     |  24%  |                                                                              |=================                                                     |  25%  |                                                                              |==================                                                    |  25%  |                                                                              |==================                                                    |  26%  |                                                                              |===================                                                   |  27%  |                                                                              |===================                                                   |  28%  |                                                                              |====================                                                  |  28%  |                                                                              |======================                                                |  32%  |                                                                              |=======================                                               |  32%  |                                                                              |=======================                                               |  33%  |                                                                              |========================                                              |  34%  |                                                                              |========================                                              |  35%  |                                                                              |=========================                                             |  35%  |                                                                              |=========================                                             |  36%  |                                                                              |==========================                                            |  37%  |                                                                              |==========================                                            |  38%  |                                                                              |===========================                                           |  38%  |                                                                              |===========================                                           |  39%  |                                                                              |============================                                          |  39%  |                                                                              |============================                                          |  40%  |                                                                              |============================                                          |  41%  |                                                                              |=============================                                         |  41%  |                                                                              |=============================                                         |  42%  |                                                                              |==============================                                        |  42%  |                                                                              |==============================                                        |  43%  |                                                                              |==============================                                        |  44%  |                                                                              |===============================                                       |  44%  |                                                                              |===============================                                       |  45%  |                                                                              |================================                                      |  45%  |                                                                              |================================                                      |  46%  |                                                                              |=================================                                     |  47%  |                                                                              |=================================                                     |  48%  |                                                                              |==================================                                    |  48%  |                                                                              |==================================                                    |  49%  |                                                                              |===================================                                   |  49%  |                                                                              |===================================                                   |  50%  |                                                                              |===================================                                   |  51%  |                                                                              |====================================                                  |  51%  |                                                                              |====================================                                  |  52%  |                                                                              |=====================================                                 |  52%  |                                                                              |=====================================                                 |  53%  |                                                                              |======================================                                |  54%  |                                                                              |======================================                                |  55%  |                                                                              |=======================================                               |  55%  |                                                                              |=======================================                               |  56%  |                                                                              |========================================                              |  57%  |                                                                              |========================================                              |  58%  |                                                                              |=========================================                             |  58%  |                                                                              |=========================================                             |  59%  |                                                                              |==========================================                            |  60%  |                                                                              |==========================================                            |  61%  |                                                                              |===========================================                           |  61%  |                                                                              |===========================================                           |  62%  |                                                                              |============================================                          |  62%  |                                                                              |============================================                          |  63%  |                                                                              |============================================                          |  64%  |                                                                              |=============================================                         |  64%  |                                                                              |=============================================                         |  65%  |                                                                              |==============================================                        |  65%  |                                                                              |==============================================                        |  66%  |                                                                              |===============================================                       |  67%  |                                                                              |===============================================                       |  68%  |                                                                              |================================================                      |  68%  |                                                                              |================================================                      |  69%  |                                                                              |=================================================                     |  69%  |                                                                              |=================================================                     |  70%  |                                                                              |==================================================                    |  71%  |                                                                              |==================================================                    |  72%  |                                                                              |===================================================                   |  72%  |                                                                              |===================================================                   |  73%  |                                                                              |====================================================                  |  74%  |                                                                              |====================================================                  |  75%  |                                                                              |=====================================================                 |  75%  |                                                                              |=====================================================                 |  76%  |                                                                              |======================================================                |  77%  |                                                                              |======================================================                |  78%  |                                                                              |=======================================================               |  78%  |                                                                              |=======================================================               |  79%  |                                                                              |========================================================              |  80%  |                                                                              |========================================================              |  81%  |                                                                              |=========================================================             |  81%  |                                                                              |=========================================================             |  82%  |                                                                              |==========================================================            |  82%  |                                                                              |==========================================================            |  83%  |                                                                              |===========================================================           |  84%  |                                                                              |===========================================================           |  85%  |                                                                              |============================================================          |  85%  |                                                                              |============================================================          |  86%  |                                                                              |=============================================================         |  87%  |                                                                              |=============================================================         |  88%  |                                                                              |==============================================================        |  88%  |                                                                              |==============================================================        |  89%  |                                                                              |===============================================================       |  90%  |                                                                              |===============================================================       |  91%  |                                                                              |================================================================      |  91%  |                                                                              |================================================================      |  92%  |                                                                              |=================================================================     |  92%  |                                                                              |=================================================================     |  93%  |                                                                              |=================================================================     |  94%  |                                                                              |==================================================================    |  94%  |                                                                              |==================================================================    |  95%  |                                                                              |===================================================================   |  95%  |                                                                              |===================================================================   |  96%  |                                                                              |====================================================================  |  97%  |                                                                              |====================================================================  |  98%  |                                                                              |===================================================================== |  98%  |                                                                              |===================================================================== |  99%  |                                                                              |======================================================================|  99%  |                                                                              |======================================================================| 100%
#> File downloaded!
#> Starting with the vectortranslate operations on the input file!
#> 0...10...20...30...40...50...60...70...80...90...100 - done.
#> Finished the vectortranslate operations on the input file!
#> Reading layer `lines' from data source 
#>   `C:\Users\user\AppData\Local\Temp\RtmpyclUjo\geofabrik_isle-of-man-140101.gpkg' 
#>   using driver `GPKG'
#> Simple feature collection with 8822 features and 10 fields
#> Geometry type: LINESTRING
#> Dimension:     XY
#> Bounding box:  xmin: -20.05167 ymin: 53.32345 xmax: -2.912515 ymax: 55.56667
#> Geodetic CRS:  WGS 84
```

<sup>Created on 2024-11-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

agila5 commented:
cc also @juanfonsecaLS1! Let me know if this implementation is ok. In that case, I will also add an example to the package's vignette (since I'm not sure how to formally create a test or an example without stressing too much Geofabrik servers).

Robinlovelace commented on R/get.R:
```diff
@@ -44,6 +44,12 @@
 #'   say that smaller administrative units correspond to bigger levels. If
 #'   `NULL`, the default, the `oe_*` functions will select the highest available
 #'   level. See Details and Examples in [oe_match()].
+#' @param version The version of the OSM extract to download. The default is
```
Great solution, and :+1: to backwards compatibility. Quick question: will this still default to *not* downloading new data if there's already (possibly out of date) files matching the region in the download directory?

Robinlovelace commented:
Another question @agila5: do we and should we start adding dates to the filenames?

agila5 commented:
> Another question @agila5: do we and should we start adding dates to the filenames?

IMO no since OSM data are updated almost continuously and saving each tiny difference detected in OSM files kinda defeats the purpose of storing the data. Anyway, feel free to create a new issue on that topic and we can discuss there!

juanfonsecaLS1 commented:
This is great! I think it is a good way to do it. Probably, the documentation of `oe_get_network` could also be updated to mention the `version` parameter. 

On the test, how about using a small file, e.g. [Monaco](https://download.geofabrik.de/europe/monaco.html#)  which is less than 600 KB?

Thanks for your work!

Robinlovelace commented:
> On the test, how about using a small file, e.g. [Monaco](https://download.geofabrik.de/europe/monaco.html#) which is less than 600 KB?

:+1: :pray:

agila5 commented:
> Fine by me, it' worth documenting the behaviour and stating that people need to save their own time-stamped extracts if they want versions, not something that the package does.

Will do as soon as I update the vignette (#236). Please notice that, after the current PR, we kinda have a versioning system as long as a user requires the adoption of an old (and fixed) OSM extract instead of the latest one (which is the default). For example, `oe_get("Italy")` points to the latest pbf file which is different and updated more or less every day. On the other hand, `oe_get("Italy", version = "200101")` points to a fixed OSM extract (which included the data up to 2020-01-01) that (AFAIK) will remain the same state forever (or as long as Geofabrik servers keep it). 

> Just to check, the behaviour is that force_download = FALSE, meaning that re-running oe_get() will not lead to a re-downloading of the data, right?

Yes. When `force_download = FALSE` (the default) then the data are not re-downloaded when the same .osm.pbf file already exists in the download directory. 

> Follow-on question: should force_download be switched to TRUE when version != "latest"?

IMO no since each fixed (and not default) OSM extract points to a different .pbf file. 

![image](https://github.com/user-attachments/assets/3b348653-a490-4f01-8927-76f131addbb8)

In fact, I see no pratical reason for re-downloading the same old pbf file since the data will be (or should be) always the same.

> This is great! I think it is a good way to do it. Probably, the documentation of oe_get_network could also be updated to mention the version parameter.

Done

> On the test, how about using a small file, e.g. [Monaco](https://download.geofabrik.de/europe/monaco.html#) which is less than 600 KB?

I'm not sure since the problems are not really related to the size of the extract, but more rate limit of Geofabrik website.... 

Please let me know if this is clear.

agila5 commented:
@Robinlovelace @juanfonsecaLS1 do you have any further comment after the last commits? Otherwise I will merge.

Robinlovelace commented:
Discussing this new functionality with @eugenividal, could be useful for his project. Heads-up Eugeni, you can install the version of this package that has the new feature with the following:

```r
pak::pak("ropensci/stats19@version")
```

agila5 commented:
Great! If you have any feedback please let me know :) Hopefully will start working again on `osmextract` in a few months 🤞

Robinlovelace commented:
Looks good, OK to merge this to main branch @agila5 ?

eugenividal commented:
Here's a rewritten version of your message:

---

Hi everyone, I’ve been exploring this very useful feature. Hope you’re all doing well, especially @agila5! 😊

I wanted to share my experience so far:

**Heads-up:** You can install the package version with the new feature using the following command:  
```R
pak::pak("ropensci/osmextract@version")
```

However, when I tried running it, I encountered this error:

```  
Error:                                                
! error in pak subprocess
Caused by error: 
! Could not solve package dependencies:
* ropensci/osmextract@version: ! pkgdepends resolution error for
ropensci/osmextract@version.
Caused by error: 
! Can't find reference @version in GitHub repo ropensci/osmextract.
```

Additionally, when I use version `0.5.2`, the feature doesn’t seem to work as expected. For example:

```R
> packageVersion("osmextract")
[1] ‘0.5.2’
> oe_match("Italy", version = "200101", quiet = TRUE)$url # OSM data up to January 1st 2020
[1] "https://download.geofabrik.de/europe/italy-latest.osm.pbf"
```

It seems I might need to update the package version. Does anyone have suggestions on how I can successfully install the version with the new feature? Thanks in advance!

agila5 commented:
Could you please run `remotes::install_github("ropensci/osmextract")` and report the output?

eugenividal commented:
```
> remotes::install_github("ropensci/osmextract")
Using GitHub PAT from the git credential store.
Downloading GitHub repo ropensci/osmextract@HEAD
Registered S3 methods overwritten by 'callr':
  method                    from
  format.callr_status_error     
  print.callr_status_error      
── R CMD build ──────────────────────────────────────
✔  checking for file ‘/tmp/RtmpJjZMIt/build-22b393bbf6639/osmextract/DESCRIPTION’ ...
─  preparing ‘osmextract’:
✔  checking DESCRIPTION meta-information ...
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
─  building ‘osmextract_0.5.2.9000.tar.gz’
   
Installing package into ‘/home/eugeni/R/x86_64-pc-linux-gnu-library/4.4’
(as ‘lib’ is unspecified)
* installing *source* package ‘osmextract’ ...
** using staged installation
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (osmextract)
```

agila5 commented:
It worked! Could you please try running the same code as the previous example, i.e. `oe_match("Italy", version = "200101", quiet = TRUE)$url` and report back? Thanks

eugenividal commented:
I did it but I got the latest version again 😕
```
> packageVersion("osmextract")
[1] ‘0.5.2.9000’
> oe_match("Italy", version = "200101", quiet = TRUE)$url # OSM data up to January 1st 2020
[1] "https://download.geofabrik.de/europe/italy-latest.osm.pbf"
```

agila5 commented:
Ouch, guess you've found a bug....

Robinlovelace commented:
:tada:
